### ISSUE TITLE
LAPACK error in `vector_transport_to` on `ProductManifold` containing empty Stiefel.

### ISSUE THREAD

**Issue by lukas-weber** [JuliaManifolds/Manifolds.jl#597]
I encountered another issue when I have a product manifold with an empty Stiefel submanifold.
```julia
M = ProductManifold(Stiefel(2,2), Stiefel(2,0))
x = rand(M)
y = rand(M)
X = project(M, x, rand(M))

vector_transport_to(M, x, X, y)
```
gives
```
 ** On entry to ZHEEVR parameter number 15 had an illegal value
ERROR: ArgumentError: invalid argument #15 to LAPACK call
Stacktrace:
  [1] chklapackerror
    @ /mnt/sw/nix/store/744skamw007p0finbnhdr4b124wb6hyi-julia-1.8.3/share/julia/stdlib/v1.8/LinearAlgebra/src/lapack.jl:39 [inlined]
  [2] syevr!(jobz::Char, range::Char, uplo::Char, A::Matrix{ComplexF64}, vl::Float64, vu::Float64, il::Int64, iu::Int64, abstol::Float64)
    @ LinearAlgebra.LAPACK /mnt/sw/nix/store/744skamw007p0finbnhdr4b124wb6hyi-julia-1.8.3/share/julia/stdlib/v1.8/LinearAlgebra/src/lapack.jl:5269
  [3] #eigen!#209
    @ /mnt/sw/nix/store/744skamw007p0finbnhdr4b124wb6hyi-julia-1.8.3/share/julia/stdlib/v1.8/LinearAlgebra/src/symmetriceigen.jl:4 [inlined]
  [4] #eigen#210
    @ /mnt/sw/nix/store/744skamw007p0finbnhdr4b124wb6hyi-julia-1.8.3/share/julia/stdlib/v1.8/LinearAlgebra/src/symmetriceigen.jl:10 [inlined]
  [5] eigen
    @ /mnt/sw/nix/store/744skamw007p0finbnhdr4b124wb6hyi-julia-1.8.3/share/julia/stdlib/v1.8/LinearAlgebra/src/symmetriceigen.jl:7 [inlined]
  [6] sqrt(A::LinearAlgebra.Hermitian{ComplexF64, Matrix{ComplexF64}}; rtol::Float64)
    @ LinearAlgebra /mnt/sw/nix/store/744skamw007p0finbnhdr4b124wb6hyi-julia-1.8.3/share/julia/stdlib/v1.8/LinearAlgebra/src/symmetric.jl:842
  [7] sqrt
    @ /mnt/sw/nix/store/744skamw007p0finbnhdr4b124wb6hyi-julia-1.8.3/share/julia/stdlib/v1.8/LinearAlgebra/src/symmetric.jl:840 [inlined]
  [8] sqrt(A::Matrix{ComplexF64})
    @ LinearAlgebra /mnt/sw/nix/store/744skamw007p0finbnhdr4b124wb6hyi-julia-1.8.3/share/julia/stdlib/v1.8/LinearAlgebra/src/dense.jl:855
  [9] vector_transport_to_diff!(M::Stiefel{2, 0, ℂ}, Y::Matrix{ComplexF64}, p::Matrix{ComplexF64}, X::Matrix{ComplexF64}, q::Matrix{ComplexF64}, #unused#::PolarRetraction)
    @ Manifolds ~/.julia/packages/Manifolds/ClaN0/src/manifolds/Stiefel.jl:700
 [10] #vector_transport_to_diff#160
    @ ~/.julia/packages/ManifoldsBase/mQYOL/src/vector_transport.jl:1169 [inlined]
 [11] vector_transport_to_diff
    @ ~/.julia/packages/ManifoldsBase/mQYOL/src/vector_transport.jl:1167 [inlined]
 [12] #_vector_transport_to#158
    @ ~/.julia/packages/ManifoldsBase/mQYOL/src/vector_transport.jl:1150 [inlined]
 [13] _vector_transport_to
    @ ~/.julia/packages/ManifoldsBase/mQYOL/src/vector_transport.jl:1142 [inlined]
 [14] vector_transport_to
    @ ~/.julia/packages/ManifoldsBase/mQYOL/src/vector_transport.jl:1130 [inlined]
 [15] vector_transport_to(::ManifoldsBase.EmptyTrait, M::Stiefel{2, 0, ℂ}, p::Matrix{ComplexF64}, X::Matrix{ComplexF64}, q::Matrix{ComplexF64}, m::DifferentiatedRetractionVectorTransport{PolarRetraction})
    @ ManifoldsBase ~/.julia/packages/ManifoldsBase/mQYOL/src/nested_trait.jl:346
```

Of course, I could just exclude that manifold, but it would be quite handy if that corner case was handled properly (which for the most part it appears to be). For example `M = ProductManifold(Stiefel(2,2), Euclidean(0))` works.

**Comment by mateuszbaran** [comment]
We don't really test such zero-dimensional cases so I wouldn't be surprised if there is more things that won't work. This looks fairly simple to fix though, I will make a PR tomorrow.

**Comment by kellertuer** [comment]
Uh, I have never seen a `Stiefel(2,0)` and would be very interested in hearing about where such a manifold is used.

**Comment by lukas-weber** [comment]
I wound up using it in physics. Basically, I am doing a [Hartree-Fock](https://en.wikipedia.org/wiki/Hartree%E2%80%93Fock_method) calculation, that is, I minimize the energy $E = ⟨\psi, H[\psi] \psi⟩$ where $\psi$ is a Slater determinant state and the Hamiltonian $H[\psi]$ is an operator on a Hilbert space with a nonlinear dependence on $\psi$. A useful representation of $\psi$ is in terms of $N$ orthonormal vectors where $N$ is the number of particles.

There is a spin rotational symmetry that makes $H$ block-diagonal in my basis, decomposing into two blocks and $\psi$ into two parts (corresponding to spin up and spin down). To use this structure, I take $\psi$ to be on the product of Stiefel manifolds $V_{N_\uparrow} \times V_{N_\downarrow}$, but in the special case $N_\downarrow = 0$, the zero-dimensional manifold appears.

I completely understand if that’s not something you want to account for.

**Comment by kellertuer** [comment]
Thanks!

That was why I stated that I am interested. Mateusz knows these details better than me so if he sees a PR to fix that – we will surely cover that :)
But an alternative is of course to just not do a product manifold here but a plain single Stiefel. But I see that when N is a parameter in you setting and you want to generically suppeort $N=0$ this naturally appears (if you do not handle that as a special case in maybe too many places).

So – for myself it is not about _want_ to support, more like can. If Mateusz sees a way – we of course _can_ support it!
Or if you see one and do a PR, as a hint for future things that _you_ might want :) (of course properly documented and tested and such, but we hapily help with any PR for sure).

**Comment by lukas-weber** [comment]
Thanks! I thought about the single Stiefel, but vectors in the different parts of $\psi$ need not be orthogonal (since they live in different blocks). Increasing the dimension of the Stiefel to account for that would be possible but not as convenient and unnecessarily adding a lot of optimization parameters.

I will keep it in mind. If there is something within my knowledge to implement properly, I will open a PR.

**Comment by kellertuer** [comment]
Oh, I was maybe not precise enough, I meant a single Stiefel _only_ for the case $N=0$, not in general – the disadvantage is then that you have too maybe treat the case $N=0$ in a lot of cases differently. But again, since Mateusz seems to have an idea, wait for his PR :)

**Comment by mateuszbaran** [comment]
Weird, I can't reproduce it now. Could you try if `sqrt(fill(1.0, 0, 0))` works on your Julia? Or if your example works on Julia 1.8.5?

**Comment by lukas-weber** [comment]
Oh my bad, I must have confused my manifolds when simplifying my example. I get the error with complex Stiefel. Correspondingly, `sqrt(fill(1im, 0, 0))` fails on julia 1.8.5 (and 1.8.3 which I used yesterday). Seems like the problem is upstream then. I will check with the most recent unstable julia and open an issue there.

Thanks for looking into it!

**Comment by mateuszbaran** [comment]
I see, the complex variant gives me your error. This looks like a Julia issue. Anyway, it's still present on Julia's master branch so I could add a workaround in Manifolds.jl

**Comment by lukas-weber** [comment]
I think if a workaround is needed anyway, I might as well add it on my side.

**Comment by kellertuer** [comment]
Is this clarified/fixed? Then we could close this issue.

**Comment by mateuszbaran** [comment]
It's fixed in Julia 1.10: https://github.com/JuliaLang/julia/issues/49667. I think we can close it.
